chain/client/src/sync/state.rs

```diff
@@ -711,27 +677,13 @@ impl StateSync {
     ) -> Result<StateSyncResult, near_chain::Error> {
         let _span = tracing::debug_span!(target: "sync", "run", sync = "StateSync").entered();
         tracing::trace!(target: "sync", %sync_hash, ?tracking_shards, "syncing state");
-        let prev_hash = *chain.get_block_header(&sync_hash)?.prev_hash();
         let now = StaticClock::utc();
 
-        // FIXME: it checks if the block exists.. but I have no idea why..
-        // seems that we don't really use this block in case of catchup - we use it only for state sync.
-        // Seems it is related to some bug with block getting orphaned after state sync? but not sure.
-        let (request_block, have_block) = self.sync_block_status(&prev_hash, chain, now)?;
-
         if tracking_shards.is_empty() {
             // This case is possible if a validator cares about the same shards in the new epoch as
             //    in the previous (or about a subset of them), return success right away
 
-            return if !have_block {
-                if request_block {
-                    Ok(StateSyncResult::RequestBlock)
-                } else {
-                    Ok(StateSyncResult::InProgress)
-                }
-            } else {
-                Ok(StateSyncResult::Completed)
-            };
+            return Ok(StateSyncResult::Completed);
```

Comment:
So is this change here the reason for including `if have_block { }` when running state sync in the client actor above? Because otherwise this part would not keep the old logic unless we guarantee that this function is only run when we do have the block? If so, it seems sort of sad to give up on requesting state sync headers before we have the block, just to be able to keep the old logic here. I think we can still get rid of the `if have_block { }` in the client actor, and then here just say:

```
return if have_block {
  Ok(StateSyncResult::Completed)
} else {
  Ok(StateSyncResult::InProgress)
}
```

Even though the `have_block` has been removed from this function, it's still easy to either pass it as an argument, or just recompute it here, since it's just a simple `chain.block_exists()` call

Reply:
What do you think about returning `Completed` here and in client actor wait for the block to be processed.

Reply:
Hm, so then we change some of the logic around when we switch to block sync? If we do what you suggest then we need to change more code in the client actor right? Because otherwise it would just switch to `BodySync` after returning from this function. Without digging deeper I'm not 100% sure what the problem with immediately switching to `BodySync` was when the current code was written, but maybe something will go wrong somewhere

 So I guess I'm not opposed to it in principle, but is there a reason we want to change this? If this is just a refactor, why not keep the old behavior here since it's pretty easy? and then we don't have to worry about introducing new bugs with a refactor

Reply:
It seems to me you run `run()` regardless of whether `run_sync_step()` has `have_block` or not.
I believe you need `have_block=true` for the transition to BlockSync, otherwise, BlockSync will trigger StateSync again, and then again,...

Reply:
In my [previous iteration](https://github.com/near/nearcore/pull/10175#discussion_r1396492271) I was not running `run` without the block but I got pushback on it.

Now it runs state sync but it does not proceed to block sync. In if state sync finishes before the block is fetched,  `Complete` case, it will not proceed to block sync.
```
                            if !have_block {
                                info!(target: "sync", "Sync done. Waiting for sync block.");
                                return;
                            }
```

Reply:
Yea I see that now after this latest change we'll just wait to proceed further after `Completed` is returned if `have_block` is false, so LGTM now. although i wont mark it as an approval from me right now because I think my other comment about the log spam is actually sort of serious/blocking even though its just a simple logging issue, because it could be pretty unpleasant. But after that is changed feel free to just merge